update/application.go

```diff
@@ -0,0 +1,105 @@
+	"fmt"
+
+	apps "github.com/ninech/apis/apps/v1alpha1"
+	"github.com/ninech/nctl/api"
+	"github.com/ninech/nctl/api/util"
+	"github.com/ninech/nctl/internal/format"
+)
+
+// all fields need to be pointers so we can detect if they have been set by
+// the user.
+type applicationCmd struct {
+	Name     *string            `arg:"" help:"Name of the application."`
+	Git      *gitConfig         `embed:"" prefix:"git-"`
+	Size     *string            `help:"Size of the app."`
+	Port     *int32             `help:"Port the app is listening on."`
+	Replicas *int32             `help:"Amount of replicas of the running app."`
+	Hosts    *[]string          `help:"Host names where the application can be accessed. If empty, the application will just be accessible on a generated host name on the deploio.app domain."`
+	Env      *map[string]string `help:"Environment variables which are passed to the app at runtime."`
+}
+
+type gitConfig struct {
+	URL           *string `help:"URL to the Git repository containing the application source. Both HTTPS and SSH formats are supported."`
+	SubPath       *string `help:"SubPath is a path in the git repo which contains the application code. If not given, the root directory of the git repo will be used."`
+	Revision      *string `help:"Revision defines the revision of the source to deploy the application to. This can be a commit, tag or branch."`
+	Username      *string `help:"Username to use when authenticating to the git repository over HTTPS." env:"GIT_USERNAME"`
+	Password      *string `help:"Password to use when authenticating to the git repository over HTTPS. In case of GitHub or GitLab, this can also be an access token." env:"GIT_PASSWORD"`
+	SSHPrivateKey *string `help:"Private key in x509 format to connect to the git repository via SSH." env:"GIT_SSH_PRIVATE_KEY"`
+}
+
+func (cmd *applicationCmd) Run(ctx context.Context, client *api.Client) error {
+	// as name is a required arg this should not actually happen when called
+	// through kong. But we still want to handle it in case this is called
+	// directly.
+	if cmd.Name == nil {
+		return fmt.Errorf("name of the app has to be set")
+	}
+
+	app := &apps.Application{}
+	if err := client.Get(ctx, client.Name(*cmd.Name), app); err != nil {
+		return err
+	}
+	cmd.applyUpdates(app)
+
+	if cmd.Git != nil {
+		auth := util.GitAuth{
+			Username:      cmd.Git.Username,
+			Password:      cmd.Git.Password,
+			SSHPrivateKey: cmd.Git.SSHPrivateKey,
+		}
+
+		if auth.Enabled() {
+			secret := auth.Secret(app.Name, client.Namespace)
+			if err := client.Get(ctx, client.Name(secret.Name), secret); err != nil {
+				return err
+			}
+
+			auth.Update(secret)
+			if err := client.Update(ctx, secret); err != nil {
+				return err
+			}
+		}
+	}
+
+	if err := client.Update(ctx, app); err != nil {
+		return err
+	}
+
+	format.PrintSuccessf("⬆️", "updated %s %q", apps.ApplicationKind, app.Name)
+
+	return nil
+}
+
+func (cmd *applicationCmd) applyUpdates(app *apps.Application) {
```

Comment:
for application create we have create/create.go with type `creator` + methods. Maybe we should do the same for update? I mean create type `updater` + methods. Just an idea 🙂

Reply:
I usually added these as soon as we had 2+ resources, that usually makes it clearer what exactly the helper will need to do. But yeah for updates it should always be roughly the same steps so I'll add something 👍

Reply:
I added an `updater` helper. It's quite basic:

* Get the current resource from the API
* Call the updateFunc where all the custom logic per resource happens
* Call update on the API
* Print success message